File: c++/cote/PGMRS/국가예산.cpp
#include <iostream>
#include <random>
#include <vector>
#include <iomanip>

#define TEST_CASES 10
#define N 10

using namespace std;

bool under_budget(double limit, vector<int>& budgets, int m){
	double total = 0;

	for(auto budget : budgets)
		total += (budget <= limit) ? budget : limit;

	return total <= m ;
}

double solve(vector<int>& budgets, int m){

	double max_budget = *max_element(begin(budgets), end(budgets));

	double lo =0, hi= max_budget;

	// lo <= x < hi
	for(int it=0; it < 100; it++){
		double mid = (lo + hi)/2;

		if(under_budget(mid, budgets, m))
			lo = mid;
		else 
			hi = mid;
		
	}

	return 	lo;
}

int main(void){
	
	default_random_engine gen(43);
	normal_distribution<float> dis2(100, 50); 

	for(int it =0; it<TEST_CASES; it++){
		vector<int> budgets;
		for(int i=0; i<N; i++)
			budgets.push_back(abs(floor(dis2(gen))));
		int m = dis2(gen) * N;

		cout << fixed << setprecision(3) << solve(budgets, m) << endl;
		cout << endl;
	}

	return 0;
}

File: javascript/cote/PGMRS/국가예산.js

const TEST_CASE = 10
const N = 10 


function* numGenerate(n) {
	while (true){
		yield Math.floor(Math.random() * n)
	}
}

function underBudget(budgets, limit, m){

	let total = 0 ;

	for(let budget of budgets){
		total += (budget < limit) ? budget : limit
	}

	return total <= m;
}

function solve(budgets, m){

	let lo = 0, hi = Math.max(...budgets);

	for(let it=0; it<100; it++) {
		let mid = (lo+hi)/2

		if (underBudget(budgets, mid, m)){
			lo = mid
		} else {
			hi = mid
		}
	}

	return lo
}


let budgetGenerator = numGenerate(100)
let mGenerator = numGenerate(100 * N * 0.8)

for (let i=0; i< TEST_CASE; i++){

	let budgets= [];


	for(let n=0; n< N; n++){
		let { value : budget } = budgetGenerator.next()
		budgets.push(budget)
	}


	let {value : m} = mGenerator.next()

	console.log(solve(budgets, m).toPrecision(3))

} 